0010-keccak-syscall.md

```diff
@@ -0,0 +1,402 @@
+name to this RFC.
+
+The program can request data from the outside world, but instead of using an
+RPC, the data will be loaded through an oracle, called the pre-image oracle. You
+can find a better description about it in the RFC [Client/Host
+integration](./0011-cli-host-integration.md).
+
+## Motivation
+
+We aim to implement the same syscalls in the MIPS zkVM than the current version
+that Cannon supports . The behavior must be the same. The VM must read and
+write to the same registers.
+
+## Detailed design
+
+Multiple general purpose registers will be used. We will give an alias for each
+of them in this document and we will refer to them by their name.
+
+<!-- https://www.dsi.unive.it/~gasparetto/materials/MIPS_Instruction_Set.pdf -->
+Here the mapping:
+- `zero` - register number 0
+- `v0` - register number 2
+- `v1` - register number 3
+- `a0` - register number 4
+- `a1` - register number 5
+- `a2` - register number 6
+- `a3` - register number 7
+
+This document will also suppose that the zkVM has some pre-defined interface to
+read and write the memory and the registers, and the details of the
+implementation and the underlying generated constraints will be omitted.
+Here provided the environment:
+<!-- TODO: add more while implementing -->
+- `set_register_value(idx, v)`: set the register number `idx` to the value `v`
+- `get_register_value(idx, buffer_v)`: copy the register number `idx` in the variable `buffer_v`
+- `set_halted(exit_code)`: constraint halting the program with the exit code `exit_code`.
+
+At the beginning of the instruction, the syscall number is in the register `v0`.
+
+The list of supported syscalls are:
+- `sysMmap`, mapped to value `4090`
+- `sysBrk`, mapped to value `4045`
+- `sysClone`, mapped to value `4120`
+- `sysExitGroup`, mapped to value `4246`
+- `sysRead`, mapped to value `4003`
+- `sysWrite`, mapped to value `4004`
+- `sysFcntl`, mapped to value `4055`
+
+Except in the case of the syscall `sysExitGroup`, at the end of the instruction,
+the instruction pointer must be increased by a word (32bits) as the next
+instruction pointer.
+
+For any other syscall value, the value `0` must be set into the registers `v0`
+and `a3`.
+
+An important note is that IO operations are truncated to `4` bytes, meaning that
+the register `v0` is at most `4`.
+Another important note is that addresses must be aligned at 4 bytes. IO syscalls
+**must** support unaligned addresses, i.e. the address `42` can be given even
+though it is not a multiple of `4`.
+Computing the effective address of an address `addr` can be done using the
+bitwise AND operator with `0xFFffFFfc`, which sets the latest two bits to `0`,
+i.e. takes the previous multiple of `4` which represents the aligned memory
+address. For instance, for the unaligned address `42`, the address `40` must be
+used while fetching the memory.
+
+Note that as the syscall instruction must be encoded in a circuit and its
```

Comment:
Mmmm so this sentence means that there will be a selector for each instruction type. So there's one for the syscall instruction. And when that selector is true, then the subtype of syscall being executed in particular is given by the content of the v0 register?